### ISSUE TITLE
Enhancement: explicit-function-return-type disallow "JSX.Element" return type

### ISSUE THREAD

**Issue by devinrhode2** [typescript-eslint/typescript-eslint#5517]
### Before You File a Proposal Please Confirm You Have Done The Following...

- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22enhancement%3A+plugin+rule+option%22) and found none that match my proposal.
- [X] I have searched the [current rule list](https://typescript-eslint.io/rules/#supported-rules) and found no rules that match my proposal.
- [X] I have [read the FAQ](https://typescript-eslint.io/docs/linting/troubleshooting) and my problem is not listed.

### My proposal is suitable for this project

- [X] I believe my proposal would be useful to the broader TypeScript community (meaning it is not a niche proposal).

### Link to the rule's documentation

https://github.com/typescript-eslint/typescript-eslint/blob/main/packages/eslint-plugin/docs/rules/explicit-function-return-type.md

### Description

Search query: `disallow "JSX.Element" return type`

See: https://kentcdodds.com/blog/how-to-write-a-react-component-in-typescript
And: https://twitter.com/nickemccurdy/status/1365384372908621833

cc @nickmccurdy

There could also be options for ReactNode, ReactElement, etc

Generally we only need a type def for the props, and can let return type be inferred, or at least not be JSX.Element

### Fail

```typescript
const foo = (): JSX.Element => <div>asdf</div>

// Unrelated to react return types, should still require explicit return type annoation:
const foo = () => {
  if (window.foo) {
    return '5' as string
  } else {
    return 5 as number
  }
}
```

### Pass

```typescript
const foo = () => <div>asdf</div>

const foo: number | string = () => {
  if (window.foo) {
    return '5' as string
  } else {
    return 5 as number
  }
}
```

**Comment by bradzacher** [comment]
Thanks for the issue!

This rule is about enforcing return type annotations for all functions.

See previous discussion about allowing the the rule to not require an annotation for `void` functions (https://github.com/typescript-eslint/typescript-eslint/issues/50).

TL;DR - we won't be adding this to the rule.

**Comment by devinrhode2** [comment]
Great reasoning you have at the end of that thread!!
